cmd/influx_inspect/export.go

```diff
@@ -1,92 +1,162 @@
-		dirs := strings.Split(relPath, string(byte(os.PathSeparator)))
-		database := dirs[0]
-		fmt.Println("Opening", f, "in database", database)
-		temp, _ := ioutil.TempDir("/tmp", "influxdb")
-
-		if err != nil {
-			fmt.Printf("%v", err)
-			os.Exit(1)
-		}
-
-		reader, err := tsm1.NewTSMReader(file)
+func (c *cmdExport) writeFiles() error {
+	// open our output file and create an output buffer
+	var w io.WriteCloser
+	w, err := os.Create(c.out)
+	if err != nil {
+		return err
+	}
+	defer w.Close()
+	if c.compress {
+		w = gzip.NewWriter(w)
+	}
 
-		if err != nil {
-			fmt.Printf("%v", err)
-			os.Exit(1)
-		}
+	// Write out all the DDL
+	fmt.Fprintln(w, "# DDL")
+	for key, _ := range c.files {
+		keys := strings.Split(key, string(byte(os.PathSeparator)))
+		fmt.Fprintf(w, "CREATE DATABASE %s\n", keys[0])
+		fmt.Fprintf(w, "CREATE RETENTION POLICY %s ON %s DURATION inf REPLICATION 1\n", keys[1], keys[0])
+	}
 
-		filename := temp + "/" + database + ".lp"
-
-		of, _ := os.Create(filename)
-		defer of.Close()
-		w := bufio.NewWriter(of)
-
-		fmt.Println("Writing", reader.KeyCount(), "series to", filename)
-
-		for i := 0; i < reader.KeyCount(); i++ {
-			var pairs string
-			key, typ := reader.KeyAt(i)
-			values, _ := reader.ReadAll(key)
-			split := strings.Split(key, "#!~#")
-			measurement, field := split[0], split[1]
-
-			for _, value := range values {
-				switch typ {
-				case tsm1.BlockFloat64:
-					pairs = field + "=" + fmt.Sprintf("%v", value.Value())
-				case tsm1.BlockInteger:
-					pairs = field + "=" + fmt.Sprintf("%vi", value.Value())
-				case tsm1.BlockBoolean:
-					pairs = field + "=" + fmt.Sprintf("%v", value.Value())
-				case tsm1.BlockString:
-					pairs = field + "=" + fmt.Sprintf("\"%v\"", value.Value())
-				default:
-					pairs = field + "=" + fmt.Sprintf("%v", value.Value())
+	fmt.Fprintln(w, "# DML")
+	for key, files := range c.files {
+		keys := strings.Split(key, string(byte(os.PathSeparator)))
+		fmt.Fprintf(w, "# CONTEXT-DATABASE:%s\n", keys[0])
+		fmt.Fprintf(w, "# CONTEXT-RETENTION-POLICY:%s\n", keys[1])
+		for _, f := range files {
+			// use an anonymous function here to close the files in the defers and not let them
+			// accumulate in the loop
+			if err := func(f string) error {
+				file, err := os.OpenFile(f, os.O_RDONLY, 0600)
+				if err != nil {
+					return fmt.Errorf("%v", err)
 				}
-
-				fmt.Fprintln(w, measurement, pairs, value.UnixNano())
+				defer file.Close()
+				reader, err := tsm1.NewTSMReader(file)
+				if err != nil {
+					log.Println("unable to read %s, skipping", file)
+					return nil
+				}
+				defer reader.Close()
+
+				for i := 0; i < reader.KeyCount(); i++ {
+					var pairs string
+					key, typ := reader.KeyAt(i)
+					values, _ := reader.ReadAll(key)
+					measurement, field := tsm1.SeriesAndFieldFromCompositeKey(key)
+
+					for _, value := range values {
+						switch typ {
+						case tsm1.BlockFloat64:
+							pairs = field + "=" + fmt.Sprintf("%v", value.Value())
```

Comment:
Will this format `float64(0)` as `0.0`?  Floats should include a decimal to ensure their types won't switch to an int inadvertently.